### ISSUE TITLE
Bug: [no-unused-vars] Decorators are flagged as unused when it's actually used since TypeScript 4.8

### ISSUE THREAD

**Issue by chenxinyanc** [typescript-eslint/typescript-eslint#5496]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.0-beta&sourceType=module&showAST=es&code=PTAEHIFkE8BEFMDGB7ATgQwC5vKAlgM6gAm8AZngHbzGgBGArpqNQG7yqgME0B0oAUQDKAGSrMAFAAFM0AA7wCiVHjmYAtIoA244JWTqGlbjXWt0qAgEoAsACgyRxJjzJKoGAhQZsqADwAKqDwAB6Y8JTERABiTi5uAHwSAPqYFgDm8JgAXFyUANb6AO6UADSgyfnw0LkEmCqU6eWkSipqaLkB8jQACqjICqiyCK2qvoEJVp3dxH0DHMOKymNoE6AA3vagoKhZDKjuLcvtqADc9gC+9vaIWugERDAAwncPG1ugUp5IaFhoH3IGHQdIhQNFkMgJFZ3nZtlc7PD7CBQAEABb9IoeaAve5EbCgTCo+D0O6IfKgVHILTwXj2VjIPDECTPV7WexAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgwBtcNhyJo0DtEgAaWXKxzskMRPRRR8eGo3bI4gOaTDAM3wLu-Aae07KlztEQAlFPgBlSgAjRiZrA3w4RHU3HCtkAElPZUQABVF8B2gmA0gAPQB9VzjIMlFYSwALfABRRWUbSE4mREgzCABfMwBdWW7uoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkeAhpgObp4VTokBMADL5ABpwNSCTyZk0PADkArqlQBhABbpoAazKVxs9EJocAHoizJ46Ig1QARNbEz089rWB17htdPGR5bcB06YALKs9AAmjvTsbsIAviCxQA

### Repro Code

```typescript
// 'MyDecorator' is defined but never used. ESLint (@typescript-eslint/no-unused-vars)
function MyDecorator<T extends Function>(_target: unknown, _key: string, descriptor: TypedPropertyDescriptor<T>): TypedPropertyDescriptor<T> {
  return descriptor;
}

class MyClass {
  @MyDecorator
  public Foo() {
  }
}

// Throw MyClass to the black hole.
void(MyClass)
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unused-vars": [
      "error",
      {
        "vars": "all",
        "args": "after-used",
        "ignoreRestSiblings": true,
        "argsIgnorePattern": "^_",
        "caughtErrors": "none"
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "es2020",
    "strictNullChecks": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true
  }
}
```

### Expected Result

`MyDecorator` is used as decorator of `MyClass.Foo`.

### Actual Result

'MyDecorator' is defined but never used. ESLint (@typescript-eslint/no-unused-vars)

### Additional Info

I've just upgraded TypeScript from 4.7.4 to 4.8.1-rc and this lint error surfaces.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.33.1` |
| `@typescript-eslint/parser`            | `5.33.1` |
| `TypeScript`                           | `4.8.1-rc` |
| `ESLint`                               | `8.22.0` |
| `node`                                 | `16.16.0` |

**Comment by bradzacher** [comment]
We don't support RC versions of TS - as outlined in the error that would be printed to your console when running the linter and as outlined in our readme:
https://github.com/typescript-eslint/typescript-eslint#supported-typescript-version

See tracking issue - https://github.com/typescript-eslint/typescript-eslint/issues/5227

**Comment by chenxinyanc** [comment]
Got it. I'll track 5227 instead. Thanks for the information!

**Comment by markandrus** [comment]
FWIW, I think that this may still be a problem after #5227 has landed. I tried TypeScript 4.8 with the latest typescript-eslint yesterday, and saw this problem. Downgrading to 4.7 fixed it.

**Comment by Sayakie** [comment]
> FWIW, I think that this may still be a problem after #5227 has landed. I tried TypeScript 4.8 with the latest typescript-eslint yesterday, and saw this problem. Downgrading to 4.7 fixed it.

I had used typescript-eslint with TypeScript 4.8 and the problem is clearly resolved since 5.36.1 was landed. Could you try to invalidate your IDE cache?

**Comment by philkonick** [comment]
> FWIW, I think that this may still be a problem after #5227 has landed. I tried TypeScript 4.8 with the latest typescript-eslint yesterday, and saw this problem. Downgrading to 4.7 fixed it.

I was having the same issue today with typescript: 4.8.4 and typescript-eslint: 5.39.0. I think I have fixed it by adding the following rules to my no-unused-vars in eslintrc, then restarting the IDE:
```
"args": "all",
"ignoreRestSiblings": "false"
```

**Comment by bradzacher** [comment]
Hey everyone - please don't comment on old, closed issues. As per our [contributing guide](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md#commenting), prefer raising a fresh issue so that you can fill out the template and provide examples and extra context.
